File: killer_sudoku/image/border_clustering.py
"""Stage 4: format-agnostic anchored border clustering.

Classifies each of the 144 inner borders as cage/non-cage without any
newspaper-specific code, using k-means clustering anchored by cage-total
cells detected in Stage 3.

The 144 borders are pre-labelled structurally into two groups:
  - 36 box boundaries (at gap indices 2 and 5: between 3x3 boxes)
  - 108 cell boundaries (all other inner edges)

Two independent 2-class k-means problems are solved, one per group.
Cage-total anchor cells resolve which cluster label corresponds to cage borders.
When no anchors are available, all borders in the group return 0.5 (uncertain).

Strip sampling follows the same [x, y]-transposed convention used in
detect_borders_peak_count and _identify_borders (see border_detection.py).
"""

from __future__ import annotations

import logging
from enum import Enum

import numpy as np
import numpy.typing as npt
from scipy.signal import find_peaks
from sklearn.cluster import KMeans  # type: ignore[import-untyped]
from sklearn.preprocessing import StandardScaler  # type: ignore[import-untyped]

from killer_sudoku.image.config import BorderClusteringConfig

_log = logging.getLogger(__name__)


class BoundaryKind(Enum):
    """Whether a border lies on a 3x3 box boundary or an ordinary cell boundary."""

    BOX = "box"
    CELL = "cell"


def boundary_kind(gap_idx: int) -> BoundaryKind:
    """Return the structural kind of a border gap.

    Box boundaries occur between the 3rd and 4th rows/columns (gap_idx=2) and
    between the 6th and 7th (gap_idx=5).  In 0-indexed 8-gap space, the
    condition is gap_idx % 3 == 2.

    Args:
        gap_idx: 0-indexed gap index (0..7) between adjacent rows or columns.

    Returns:
        BoundaryKind.BOX if the gap lies on a 3x3 box boundary, else
        BoundaryKind.CELL.
    """
    return BoundaryKind.BOX if gap_idx % 3 == 2 else BoundaryKind.CELL


def strip_features(strip: npt.NDArray[np.uint8]) -> npt.NDArray[np.float64]:
    """Extract 4 summary features from a 1D min-projected border strip.

    Features: [peak_count, mean_brightness, variance, amplitude].
    Dark features (cage borders, grid lines) appear as dips in the strip; the
    inverted strip (~strip) is used for peak finding so dark dips become peaks.

    Args:
        strip: 1D array of uint8 pixel values from the border region.

    Returns:
        float64 array of shape (4,).
    """
    f: npt.NDArray[np.float64] = strip.astype(np.float64)
    inverted: npt.NDArray[np.float64] = 255.0 - f
    peaks, _ = find_peaks(inverted, height=32.0)
    return np.array(
        [
            float(len(peaks)),
            float(np.mean(f)),
            float(np.var(f)),
            float(np.max(f)) - float(np.min(f)),
        ],
        dtype=np.float64,
    )


def _sample_strip(
    warped_gry: npt.NDArray[np.uint8],
    is_horizontal: bool,
    gap_idx: int,
    along_idx: int,
    subres: int,
    sample_half: int,
    sample_margin_px: int,
) -> npt.NDArray[np.uint8]:
    """Sample a 1D min-projected strip for one interior border edge.

    Follows the sampling convention established in detect_borders_peak_count:
    the first numpy axis is treated as the x (column) direction, consistent
    with the transposed indexing used throughout the border detection code.

    Args:
        warped_gry: Warped grayscale image (resolution x resolution).
        is_horizontal: True for a horizontal border (constant y), False for
            vertical (constant x).
        gap_idx: 0-indexed gap position (0..7) in the perpendicular direction.
        along_idx: 0-indexed cell position (0..8) along the border.
        subres: Pixels per cell side.
        sample_half: Half-width of the sampling strip in pixels.
        sample_margin_px: Pixels removed from each end of the strip.

    Returns:
        1D uint8 array (the min-projected strip).
    """
    xm = ((2 * along_idx + 1) * subres) // 2
    xb = xm + sample_margin_px
    xt = xm + sample_half - sample_margin_px
    yl = ((gap_idx + 1) * subres) - sample_half
    yr = ((gap_idx + 1) * subres) + sample_half

    if is_horizontal:
        return np.asarray(np.min(warped_gry[xb:xt, yl:yr], axis=0), dtype=np.uint8)
    return np.asarray(np.min(warped_gry[yl:yr, xb:xt], axis=1), dtype=np.uint8)


def _anchor_set(
    cage_total_confidence: npt.NDArray[np.float64],
    threshold: float,
) -> set[tuple[bool, int, int]]:
    """Return the set of anchor borders from high-confidence cage-total cells.

    For a cage-total cell at (row, col), the borders above it (horizontal,
    gap_idx=row-1) and to its left (vertical, gap_idx=col-1) are cage borders.
    Outer-edge borders (row=0 or col=0) have no inner border above/left and
    are skipped.

    Returns:
        Set of (is_horizontal, gap_idx, along_idx) triples.
    """
    anchors: set[tuple[bool, int, int]] = set()
    for row in range(9):
        for col in range(9):
            if cage_total_confidence[row, col] >= threshold:
                if row > 0:
                    anchors.add((True, row - 1, col))
                if col > 0:
                    anchors.add((False, row, col - 1))
    return anchors


def _cluster_group(
    features: list[npt.NDArray[np.float64]],
    anchor_positions: list[int],
) -> npt.NDArray[np.float64]:
    """Cluster one group of border strips into cage/non-cage.

    Uses KMeans(k=2); anchor positions resolve which cluster label is cage.
    Returns 0.5 for all strips when no anchors are provided (polarity unknown).

    Args:
        features: List of 4-element feature vectors, one per border in the group.
        anchor_positions: Indices into features of known cage-border strips.

    Returns:
        float64 array of length len(features) with values in {0.0, 0.5, 1.0}.
    """
    n = len(features)
    if n == 0:
        return np.array([], dtype=np.float64)
    if not anchor_positions:
        return np.full(n, 0.5, dtype=np.float64)

    x: npt.NDArray[np.float64] = np.stack(features)
    scaler: StandardScaler = StandardScaler()
    x_scaled: npt.NDArray[np.float64] = np.asarray(
        scaler.fit_transform(x), dtype=np.float64
    )

    km: KMeans = KMeans(n_clusters=2, n_init=10, random_state=42)
    labels: npt.NDArray[np.intp] = np.asarray(km.fit_predict(x_scaled), dtype=np.intp)

    anchor_labels = labels[anchor_positions]
    counts = np.bincount(anchor_labels, minlength=2)
    cage_cluster = int(np.argmax(counts))
    _log.debug(
        "cluster_group: %d anchors, cage_cluster=%d, counts=%s",
        len(anchor_positions),
        cage_cluster,
        counts.tolist(),
    )

    cage_mask: npt.NDArray[np.bool_] = np.asarray(
        labels == cage_cluster, dtype=np.bool_
    )
    return cage_mask.astype(np.float64)


def cluster_borders(
    warped_gry: npt.NDArray[np.uint8],
    cage_total_confidence: npt.NDArray[np.float64],
    subres: int,
    config: BorderClusteringConfig,
    anchor_confidence_threshold: float = 0.5,
) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.float64]]:
    """Classify all 144 inner borders as cage/non-cage without format-specific code.

    Extracts features from each border strip, groups by BoundaryKind, uses
    cage-total anchors to resolve cluster polarity, and returns soft
    cage-border probabilities.

    Args:
        warped_gry: Perspective-corrected grayscale image (resolution x resolution).
        cage_total_confidence: Shape (9, 9) array from scan_cells Stage 3.
        subres: Pixels per cell side.
        config: Clustering parameters.
        anchor_confidence_threshold: Minimum cage_total_confidence for a cell to
            contribute positive border anchors.  Callers should pass
            CellScanConfig.anchor_confidence_threshold; defaults to 0.5.

    Returns:
        (border_x_prob, border_y_prob) — shapes (9, 8) and (8, 9) with values in
        [0, 1].  Values > 0.5 indicate a likely cage border; 0.5 means uncertain.
    """
    sample_half = subres // config.sample_fraction
    sample_margin_px = subres // config.sample_margin
    anchors = _anchor_set(cage_total_confidence, anchor_confidence_threshold)

    # Collect features grouped by BoundaryKind only — horizontal and vertical
    # borders of the same structural type are clustered together.  This doubles
    # the anchor count per group (anchors from both axes vote on the same
    # polarity decision) and halves the number of independent polarity choices.
    # Key: BoundaryKind
    # Value: list of (is_horizontal, gap_idx, along_idx, feature_vector)
    _edge_entry = tuple[bool, int, int, npt.NDArray[np.float64]]
    groups: dict[BoundaryKind, list[_edge_entry]] = {
        BoundaryKind.CELL: [],
        BoundaryKind.BOX: [],
    }

    for gap_idx in range(8):
        kind = boundary_kind(gap_idx)
        for along_idx in range(9):
            for is_h in (True, False):
                strip = _sample_strip(
                    warped_gry,
                    is_h,
                    gap_idx,
                    along_idx,
                    subres,
                    sample_half,
                    sample_margin_px,
                )
                feat = strip_features(strip)
                groups[kind].append((is_h, gap_idx, along_idx, feat))

    border_x_prob = np.zeros((9, 8), dtype=np.float64)
    border_y_prob = np.zeros((8, 9), dtype=np.float64)

    for _kind, entries in groups.items():
        features = [e[3] for e in entries]
        anchor_pos = [
            i for i, (is_h, g, a, _) in enumerate(entries) if (is_h, g, a) in anchors
        ]
        probs = _cluster_group(features, anchor_pos)
        for i, (is_h, gap_idx, along_idx, _) in enumerate(entries):
            if is_h:
                border_x_prob[along_idx, gap_idx] = probs[i]
            else:
                border_y_prob[gap_idx, along_idx] = probs[i]

    return border_x_prob, border_y_prob

File: web/src/image/borderClustering.ts
/**
 * Format-agnostic anchored border clustering: Stage 4 of the image pipeline.
 *
 * Mirrors Python's `killer_sudoku.image.border_clustering` module.
 *
 * Classifies all 144 inner borders as cage/non-cage using per-image k-means
 * clustering, anchored by cage-total cells detected in Stage 3.
 *
 * sklearn replacements:
 *   StandardScaler  → standardScale() (~15 lines)
 *   KMeans(k=2, n_init=10) → kmeans2() (~40 lines)
 *
 * Image convention: warpedGry is a standard row-major grayscale image where
 * pixel(row, col) = data[row * size + col].  (The Python source uses a
 * column-first convention that is geometrically equivalent for a square grid.)
 */

import type { BorderClusteringConfig } from './config.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Standard row-major square grayscale image: pixel(row, col) = data[row * size + col]. */
export interface GrayImage {
  readonly data: Uint8Array;
  readonly size: number; // width = height = resolution
}

/** Whether a border lies on a 3×3 box boundary or an ordinary cell boundary. */
export const enum BoundaryKind {
  BOX = 'box',
  CELL = 'cell',
}

// ---------------------------------------------------------------------------
// Public functions
// ---------------------------------------------------------------------------

/**
 * Return the structural kind of a border gap.
 *
 * Box boundaries occur between the 3rd and 4th rows/columns (gapIdx=2) and
 * between the 6th and 7th (gapIdx=5). In 0-indexed 8-gap space, the
 * condition is gapIdx % 3 === 2.
 *
 * @param gapIdx - 0-indexed gap index (0..7) between adjacent rows or columns.
 */
export function boundaryKind(gapIdx: number): BoundaryKind {
  return gapIdx % 3 === 2 ? BoundaryKind.BOX : BoundaryKind.CELL;
}

/**
 * Extract 4 position-independent features from a 1-D min-projected border strip.
 *
 * Uses order statistics (percentiles) so features depend only on the
 * *distribution* of brightness values, not where in the strip the ink sits.
 *
 * Features: [p5, p25, p50, mean].
 *   p5  (≈ minimum): how dark is the darkest part?  Cage border → ~50–150.
 *   p25: how broad is the dark region?
 *   p50 (median): is most of the strip dark or light?
 *   mean: overall ink level.
 *
 * @param strip - 1-D array of uint8 pixel values from the border region.
 * @returns Float64 array of shape (4,).
 */
export function stripFeatures(strip: Uint8Array): [number, number, number, number] {
  const sorted = Float64Array.from(strip).sort();
  const n = sorted.length;
  return [
    percentile(sorted, 5),
    percentile(sorted, 25),
    percentile(sorted, 50),
    mean(sorted, n),
  ];
}

/**
 * Classify all 144 inner borders as cage/non-cage without format-specific code.
 *
 * Extracts features from each border strip, groups by BoundaryKind, uses
 * cage-total anchors to resolve cluster polarity, and returns soft
 * cage-border probabilities.
 *
 * @param warpedGry - Perspective-corrected grayscale image (standard row-major).
 * @param cageTotalConfidence - Shape (9, 9) [row][col] confidence array from Stage 3.
 * @param subresParam - Pixels per cell side.
 * @param config - Clustering parameters.
 * @param anchorConfidenceThreshold - Minimum confidence for a cell to contribute anchors.
 * @returns [borderXProb, borderYProb] — shapes (9, 8) [col][rowGap] and (8, 9)
 *   [colGap][row] with values in [0, 1]. Values > 0.5 indicate a likely cage
 *   border; 0.5 means uncertain.
 */
export function clusterBorders(
  warpedGry: GrayImage,
  cageTotalConfidence: number[][],
  subresParam: number,
  config: BorderClusteringConfig,
  anchorConfidenceThreshold: number = 0.5,
): [number[][], number[][]] {
  const sampleHalf = (subresParam / config.sampleFraction) | 0;
  const sampleMarginPx = (subresParam / config.sampleMargin) | 0;
  const anchors = anchorSet(cageTotalConfidence, anchorConfidenceThreshold);

  // Group borders by BoundaryKind only — horizontal and vertical borders of
  // the same structural type are clustered together.  This doubles the anchor
  // count per group and halves the number of polarity choices.
  type EdgeEntry = [isH: boolean, gapIdx: number, alongIdx: number, feat: [number, number, number, number]];
  const groups = new Map<BoundaryKind, EdgeEntry[]>([
    [BoundaryKind.CELL, []],
    [BoundaryKind.BOX, []],
  ]);

  for (let gapIdx = 0; gapIdx < 8; gapIdx++) {
    const kind = boundaryKind(gapIdx);
    const group = groups.get(kind)!;
    for (let alongIdx = 0; alongIdx < 9; alongIdx++) {
      for (const isH of [true, false] as const) {
        const strip = sampleStrip(
          warpedGry,
          isH,
          gapIdx,
          alongIdx,
          subresParam,
          sampleHalf,
          sampleMarginPx,
        );
        const feat = stripFeatures(strip);
        group.push([isH, gapIdx, alongIdx, feat]);
      }
    }
  }

  // (9, 8) [col][rowGap] and (8, 9) [colGap][row]
  const borderXProb: number[][] = Array.from({ length: 9 }, () => new Array<number>(8).fill(0));
  const borderYProb: number[][] = Array.from({ length: 8 }, () => new Array<number>(9).fill(0));

  for (const [, entries] of groups) {
    const features = entries.map(e => e[3]);
    const anchorPos = entries
      .map((e, i) => ({ i, key: anchorKey(e[0], e[1], e[2]) }))
      .filter(({ key }) => anchors.has(key))
      .map(({ i }) => i);

    const probs = clusterGroup(features, anchorPos);

    for (let i = 0; i < entries.length; i++) {
      const [isH, gapIdx, alongIdx] = entries[i]!;
      if (isH) {
        // borderXProb[col][rowGap]: isH=true, gapIdx=rowGap, alongIdx=col
        borderXProb[alongIdx]![gapIdx] = probs[i]!;
      } else {
        // borderYProb[colGap][row]: isH=false, gapIdx=colGap, alongIdx=row
        borderYProb[gapIdx]![alongIdx] = probs[i]!;
      }
    }
  }

  return [borderXProb, borderYProb];
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Encode an anchor triple as a string key for Set membership. */
function anchorKey(isH: boolean, gapIdx: number, alongIdx: number): string {
  return `${isH ? 'h' : 'v'},${gapIdx},${alongIdx}`;
}

/**
 * Return the set of anchor border keys from high-confidence cage-total cells.
 *
 * For a cage-total cell at (row, col), the borders above it (horizontal,
 * gapIdx=row-1) and to its left (vertical, gapIdx=col-1) are cage borders.
 * Outer-edge borders (row=0 or col=0) have no inner border above/left.
 *
 * @param cageTotalConfidence - (9, 9) [row][col] confidence array.
 * @param threshold - Minimum confidence to use a cell as an anchor.
 */
function anchorSet(cageTotalConfidence: number[][], threshold: number): Set<string> {
  const anchors = new Set<string>();
  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      if (cageTotalConfidence[row]![col]! >= threshold) {
        if (row > 0) {
          // Horizontal border above: isH=true, gapIdx=row-1, alongIdx=col
          anchors.add(anchorKey(true, row - 1, col));
        }
        if (col > 0) {
          // Vertical border to the left: isH=false, gapIdx=col-1, alongIdx=row
          anchors.add(anchorKey(false, col - 1, row));
        }
      }
    }
  }
  return anchors;
}

/**
 * Sample a 1-D min-projected strip for one interior border edge.
 *
 * Uses standard row-major indexing: pixel(row, col) = data[row * size + col].
 *
 * For a horizontal border (between rows gap_idx and gap_idx+1):
 *   - Perpendicular direction = rows, range = boundary ± sampleHalf.
 *   - Along direction = cols, sampled from centre of alongIdx cell ± margin.
 *   - Min projected over columns → 1-D array of length 2*sampleHalf in row direction.
 *
 * For a vertical border (between cols gap_idx and gap_idx+1):
 *   - Perpendicular direction = cols, range = boundary ± sampleHalf.
 *   - Along direction = rows, sampled from centre of alongIdx cell ± margin.
 *   - Min projected over rows → 1-D array of length 2*sampleHalf in col direction.
 *
 * @param warpedGry - Standard row-major square grayscale image.
 * @param isHorizontal - True for horizontal border (between rows), false for vertical.
 * @param gapIdx - 0-indexed gap position (0..7) perpendicular to the border.
 * @param alongIdx - 0-indexed cell position (0..8) along the border.
 * @param subresParam - Pixels per cell side.
 * @param sampleHalf - Half-width of the sampling strip in pixels.
 * @param sampleMarginPx - Pixels removed from each end along the border.
 * @returns 1-D Uint8Array of min-projected pixel values, length 2*sampleHalf.
 */
function sampleStrip(
  warpedGry: GrayImage,
  isHorizontal: boolean,
  gapIdx: number,
  alongIdx: number,
  subresParam: number,
  sampleHalf: number,
  sampleMarginPx: number,
): Uint8Array {
  const { data, size } = warpedGry;
  // Centre of the cell in the "along" direction.
  const cm = (((2 * alongIdx + 1) * subresParam) / 2) | 0;
  const cStart = cm + sampleMarginPx;
  const cEnd = cm + sampleHalf - sampleMarginPx;
  // Boundary position in the perpendicular direction.
  const bndry = (gapIdx + 1) * subresParam;
  const pStart = bndry - sampleHalf;
  const pEnd = bndry + sampleHalf;
  const len = pEnd - pStart; // = 2 * sampleHalf

  const result = new Uint8Array(len).fill(255);

  if (isHorizontal) {
    // Min over columns in [cStart, cEnd), result indexed by row offset from pStart.
    for (let row = pStart; row < pEnd; row++) {
      for (let col = cStart; col < cEnd; col++) {
        const v = data[row * size + col]!;
        if (v < result[row - pStart]!) result[row - pStart] = v;
      }
    }
  } else {
    // Min over rows in [cStart, cEnd), result indexed by col offset from pStart.
    for (let col = pStart; col < pEnd; col++) {
      for (let row = cStart; row < cEnd; row++) {
        const v = data[row * size + col]!;
        if (v < result[col - pStart]!) result[col - pStart] = v;
      }
    }
  }

  return result;
}

/**
 * Cluster one group of border strips into cage/non-cage.
 *
 * Uses k-means(k=2, n_init=10); anchor positions resolve which cluster label
 * is "cage". Returns 0.5 for all strips when no anchors are provided.
 *
 * @param features - List of 4-element feature vectors.
 * @param anchorPositions - Indices into features of known cage-border strips.
 * @returns Array of length features.length with values in {0.0, 0.5, 1.0}.
 */
function clusterGroup(
  features: Array<[number, number, number, number]>,
  anchorPositions: number[],
): number[] {
  const n = features.length;
  if (n === 0) return [];
  if (anchorPositions.length === 0) return new Array<number>(n).fill(0.5);

  const xScaled = standardScale(features);
  const labels = kmeans2(xScaled, 10);

  // Resolve polarity: whichever cluster label appears most among anchors is "cage".
  const counts = [0, 0];
  for (const idx of anchorPositions) {
    counts[labels[idx]!] = counts[labels[idx]!]! + 1;
  }
  const cageCluster = counts[0]! >= counts[1]! ? 0 : 1;

  return labels.map(l => (l === cageCluster ? 1.0 : 0.0));
}

// ---------------------------------------------------------------------------
// sklearn replacements
// ---------------------------------------------------------------------------

/**
 * Zero-mean unit-variance scaling per feature column.
 * Replaces sklearn's StandardScaler.fit_transform(X).
 */
function standardScale(X: Array<[number, number, number, number]>): number[][] {
  const n = X.length;
  const d = 4;
  const means = [0, 0, 0, 0];
  const stds = [0, 0, 0, 0];

  for (const row of X) {
    for (let j = 0; j < d; j++) means[j] = means[j]! + row[j]!;
  }
  for (let j = 0; j < d; j++) means[j] = means[j]! / n;

  for (const row of X) {
    for (let j = 0; j < d; j++) {
      const diff = row[j]! - means[j]!;
      stds[j] = stds[j]! + diff * diff;
    }
  }
  // Avoid division by zero for constant features (std=0 → leave as-is).
  for (let j = 0; j < d; j++) {
    stds[j] = Math.sqrt(stds[j]! / n) || 1;
  }

  return X.map(row => row.map((v, j) => (v - means[j]!) / stds[j]!));
}

/**
 * K-means clustering with k=2 and nInit random restarts.
 * Replaces sklearn's KMeans(n_clusters=2, n_init=nInit, random_state=42).
 *
 * Returns labels array (0 or 1 per point) from the best run (lowest inertia).
 */
function kmeans2(X: number[][], nInit: number): number[] {
  const n = X.length;
  const d = X[0]!.length;
  let bestLabels: number[] = new Array<number>(n).fill(0);
  let bestInertia = Infinity;

  for (let init = 0; init < nInit; init++) {
    // Pick 2 distinct random points as initial centroids.
    const i0 = (Math.random() * n) | 0;
    let i1 = (Math.random() * (n - 1)) | 0;
    if (i1 >= i0) i1++;
    const centroids = [X[i0]!.slice(), X[i1]!.slice()];
    let labels = new Array<number>(n).fill(0);

    for (let iter = 0; iter < 100; iter++) {
      // Assign each point to the nearest centroid.
      const newLabels = X.map(p => dist2(p, centroids[0]!) <= dist2(p, centroids[1]!) ? 0 : 1);

      // Check for convergence.
      let changed = false;
      for (let i = 0; i < n; i++) {
        if (newLabels[i]! !== labels[i]!) { changed = true; break; }
      }
      labels = newLabels;
      if (!changed) break;

      // Update centroid positions.
      for (let k = 0; k < 2; k++) {
        const cnt = new Array<number>(d).fill(0);
        let sz = 0;
        for (let i = 0; i < n; i++) {
          if (labels[i]! === k) {
            for (let j = 0; j < d; j++) cnt[j] = cnt[j]! + X[i]![j]!;
            sz++;
          }
        }
        if (sz > 0) {
          for (let j = 0; j < d; j++) centroids[k]![j] = cnt[j]! / sz;
        }
      }
    }

    // Compute inertia for this run.
    let inertia = 0;
    for (let i = 0; i < n; i++) {
      inertia += dist2(X[i]!, centroids[labels[i]!]!);
    }
    if (inertia < bestInertia) {
      bestInertia = inertia;
      bestLabels = labels.slice();
    }
  }

  return bestLabels;
}

/** Squared Euclidean distance between two vectors. */
function dist2(a: number[], b: number[]): number {
  let s = 0;
  for (let i = 0; i < a.length; i++) {
    const d = a[i]! - b[i]!;
    s += d * d;
  }
  return s;
}

// ---------------------------------------------------------------------------
// Statistical helpers
// ---------------------------------------------------------------------------

/** Compute a percentile of a pre-sorted Float64Array. */
function percentile(sorted: Float64Array, p: number): number {
  const n = sorted.length;
  if (n === 0) return 0;
  const idx = (p / 100) * (n - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  return sorted[lo]! + (sorted[hi]! - sorted[lo]!) * (idx - lo);
}

/** Compute mean of a typed array. */
function mean(arr: Float64Array, n: number): number {
  let s = 0;
  for (let i = 0; i < n; i++) s += arr[i]!;
  return s / n;
}
